proj/src/main/webapp/test/chat.test.js

```diff
@@ -0,0 +1,158 @@
+// Chai is a commonly used library for creating unit test suites. It is easily extended with plugins.
+const chai = require('chai');
+const assert = chai.assert;
+
+// Firebase for setting up database for tests
+const firebase = require('firebase');
+
+// For Test Clean Up
+const test = require('firebase-functions-test')();
+
+// Set up the app
+const appconfig = require("../appconfig.js");
+const testconfig = require("./test-appconfig.js");
+appconfig.firebaseConfig = testconfig.firebaseTestConfig
+const app = require("../app.js");
+
+let chat = require('../chat.js');
+
+async function createUserIfNotExisting(userData){
+    const testEmail = userData.email
+    const testPassword = userData.password
+    // Try to sign in to user
+    userData.uid = await firebase.auth().signInWithEmailAndPassword(testEmail, testPassword).then(function(data){
+            return firebase.auth().currentUser.uid;
+        }).catch(async function(){
+            // Create the test user if it doesn't exists
+            return firebase.auth().createUserWithEmailAndPassword(testEmail, testPassword).then(function(data){
+                return firebase.auth().currentUser.uid;
+            }).catch(function(err){
+                    throw err
+                })
+            })    
+}
+
+async function deleteUser(userData){
+    // Try to sign in to user
+    await firebase.auth().signInWithEmailAndPassword(userData.email, userData.password).then(async function(data){
+        await firebase.auth().currentUser.delete()
+        }).catch(function(err){
+            })
+}
+
+describe('Creating and Joining Chats', function() {
+    this.timeout(10000);
+    let testUsers = [
+        {email:'test@test.com', password:'test123'},
+        {email:'test2@test.com', password:'test123'}]
+
+    before(async function() {
+        // Require sample.js and save the exports inside a namespace called myFunctions.
+        for (let i=0;i<testUsers.length;i++){
+            await createUserIfNotExisting(testUsers[i])
+            await firebase.auth().signOut()
+        }
+    });
+
+    after(async function() {
+        if (testconfig.shouldCleanUp) {
+            // // Delete the test users
+            // for (let i=0;i<testUsers.length;i++){
+            //     await deleteUser(testUsers[i])
+            // }
```

Comment:
Considering if we should just never delete some test users so they can be reused. If you delete and recreate users too many times, firebase starts blocking your IP.